### ISSUE TITLE
Regression: `Task.execute()` fails with NPE when task is constructed via `ProjectBuilder`

### ISSUE THREAD

**Issue by bigdaz** [gradle/gradle#5396]
`ProjectBuilder` can be used to construct task instances for unit testing. `Task.execute()` is deprecated, but should continue to function correctly until removed in a future release.

### Expected Behavior
Calling `Task.execute()` on a task instance created by `ProjectBuilder` should succeed.
This is the case for Gradle 4.7.

### Current Behavior
With Gradle 4.8 (candidate), `Task.execute` fails with a NPE.

```
nebula.plugin.dependencylock.tasks.GenerateLockTaskSpec > simple transitive lock FAILED
    java.lang.NullPointerException
        at org.gradle.internal.progress.BuildProgressLogger.afterExecute(BuildProgressLogger.java:79)
        at org.gradle.internal.progress.BuildProgressLogger.afterNestedExecute(BuildProgressLogger.java:84)
        at org.gradle.internal.progress.BuildProgressFilter.afterExecute(BuildProgressFilter.java:102)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:370)
        at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:352)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:340)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:327)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:340)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:327)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:340)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:327)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:340)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:327)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:57)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.api.internal.AbstractTask.execute(AbstractTask.java:398)
        at nebula.plugin.dependencylock.tasks.GenerateLockTaskSpec.simple transitive lock(GenerateLockTaskSpec.groovy:341)
```

### Steps to Reproduce (for bugs)

Change `task.doStuff()` to `task.execute()` here: https://github.com/gradle/gradle/blob/87bd495391cc06bfc666c20db3eae806daf019d9/subprojects/core/src/test/groovy/org/gradle/testfixtures/ProjectBuilderTest.groovy#L112

Alternatively, execute `./gradlew test` on https://github.com/nebula-plugins/gradle-dependency-lock-plugin.

### Your Environment
 * Build scan URL: https://e.grdev.net/s/jx2otmlnwftdw/

**Comment by bigdaz** [comment]
@wolfs  I suspect your refactoring to extract `EventFiringTaskExecuter` was the cause of this :point_up: . 

@gradle/build-tool  I've raised this against `4.8-RC-1` since it's a regression from 4.7. However, it's relatively minor and we might decide not to fix for 4.8 (or at all). I'm not sure how widespread this type of usage is.

**Comment by lptr** [comment]
Huh, I didn't realize `ProjectBuilder` was public API...

**Comment by bigdaz** [comment]
Yeah, neither did I until I was recently chastised for not adding test coverage when I made a change!

**Comment by lptr** [comment]
Maybe we should retire it in favor of TestKit tests?

**Comment by ldaley** [comment]
I think it is in our best interest to stop allowing people to put parts of the build machinery in motion with ProjectBuilder, as it only barely works and works by accident. As we change the execution mechanics, it will be more of a problem. So +1 on retiring.

However, I know from experience that there are quite some tests in the wild based on ProjectBuilder that do things like Task.execute(). We will need a proper deprecation cycle, and replacements (which should be TestKit).

**Comment by wolfs** [comment]
Fixed via #5406.
